Bug 1846093: Improve bootstrap reliability on heterogeneous UPI network configurations

ironcladlou opened the pull request:
Before this change, bootstrap IP discovery assumed that the first address of the
unicast interface must be the bootstrap IP. This assumption doesn't always hold
in the face of user-defined interfaces and addresses whose ordering isn't
guaranteed. When the assumptions are broken and the incorrect bootstrap IP is
selected, bootstrapping fails because quorum cannot be established.

This change improves the accuracy of bootstrap IP discovery by more flexibly
accounting for a wider variety of possible network interface configurations.

An IP is now considered the bootstrap IP if all of the following are true.

For IPv4:

* The IP is contained by the machine CIDR defined in the cluster configuration
* On bare metal platforms, the IP is not the API or DNS VIP in the cluster configuration

For IPv6, the same must be true in addition to the following:

* The IP is not deprecated
* The IP is routable according at least one non-default route

This work is adapted from https://github.com/openshift/baremetal-runtimecfg/blob/master/pkg/utils/utils.go.

ironcladlou commented:
Still need to play with this in a new cluster's bootstrapping environment, here's how it looks on an already solvent master ipv4 GCP master node:

```
I0622 20:29:00.242715      81 bootstrap_ip_linux.go:37] retrieved Address map map[0xc000262c30:[127.0.0.1/8 lo ::1/128] 0xc0005707e0:[10.129.0.74/23 eth0 fe80::7c73:6eff:fe59:f9ad/64]]
I0622 20:29:00.242841      81 bootstrap_ip_linux.go:114] Checking whether address 127.0.0.1/8 lo contains VIP 10.129.0.74
I0622 20:29:00.242929      81 bootstrap_ip_linux.go:56] Ignoring route non Router advertisement route {Ifindex: 3 Dst: fe80::/64 Src: <nil> Gw: <nil> Flags: [] Table: 254}
I0622 20:29:00.242953      81 bootstrap_ip_linux.go:66] Retrieved route map map[]
I0622 20:29:00.242976      81 bootstrap_ip_linux.go:114] Checking whether address 10.129.0.74/23 eth0 contains VIP 10.129.0.74
I0622 20:29:00.242986      81 bootstrap_ip_linux.go:116] Address 10.129.0.74/23 eth0 contains VIP 10.129.0.74
I0622 20:29:00.242995      81 bootstrap_ip_linux.go:114] Checking whether address fe80::7c73:6eff:fe59:f9ad/64 contains VIP 10.129.0.74
10.129.0.74
```

ironcladlou commented:
What role should machineCIDR play in this discovery?

ironcladlou commented:
@hexfusion @celebdor the latest commit here refactors the detection to add an additional filter that ensures candidate IPs are contained in the machine CIDR.

One caveat here is that I'm not sure if machineCIDR is always present. In the current code, I'm assuming that if machineCIDR is unavailable, no machine CIDR containment check will happen (but we do now produce a warning). The net effect in that case is that it's theoretically possible to select an IP which is routable but outside the unknown machine CIDR. I'm not entirely sure what the right assumption here should be. 

Is machine CIDR truly optional? How should the product react when machine CIDR is unknown here?

hexfusion commented:
great questions and ones that needs to be answered before this merges

hardys commented:
> @hexfusion @celebdor the latest commit here refactors the detection to add an additional filter that ensures candidate IPs are contained in the machine CIDR.
> 
> One caveat here is that I'm not sure if machineCIDR is always present. In the current code, I'm assuming that if machineCIDR is unavailable, no machine CIDR containment check will happen (but we do now produce a warning). The net effect in that case is that it's theoretically possible to select an IP which is routable but outside the unknown machine CIDR. I'm not entirely sure what the right assumption here should be.
> 
> Is machine CIDR truly optional? How should the product react when machine CIDR is unknown here?

Looking at the installer interfaces I think it's mandatory:

https://github.com/openshift/installer/blob/master/pkg/types/validation/installconfig.go#L249

It's marked as optional in the install-config schema, but AFAICS that's only to enable backwards compatibility with machineNetwork and the now-deprecated machineCIDR interface

https://github.com/openshift/installer/blob/master/pkg/types/installconfig.go#L213

https://github.com/openshift/installer/blob/master/pkg/types/installconfig.go#L234

ironcladlou commented:
@hardys 


> > @hexfusion @celebdor the latest commit here refactors the detection to add an additional filter that ensures candidate IPs are contained in the machine CIDR.
> > One caveat here is that I'm not sure if machineCIDR is always present. In the current code, I'm assuming that if machineCIDR is unavailable, no machine CIDR containment check will happen (but we do now produce a warning). The net effect in that case is that it's theoretically possible to select an IP which is routable but outside the unknown machine CIDR. I'm not entirely sure what the right assumption here should be.
> > Is machine CIDR truly optional? How should the product react when machine CIDR is unknown here?
> 
> Looking at the installer interfaces I think it's mandatory:
> 
> https://github.com/openshift/installer/blob/master/pkg/types/validation/installconfig.go#L249
> 
> It's marked as optional in the install-config schema, but AFAICS that's only to enable backwards compatibility with machineNetwork and the now-deprecated machineCIDR interface
> 
> https://github.com/openshift/installer/blob/master/pkg/types/installconfig.go#L213
> 
> https://github.com/openshift/installer/blob/master/pkg/types/installconfig.go#L234

Thank you for digging that up. Would it make sense then for us to fail fast and loudly if the machine CIDR is empty?

ironcladlou commented:
/retest

ironcladlou commented:
Latest revision disentangles render from all bootstrap IP determination by moving all that stuff into the bootstrap_* files. Also cleaned up that code and added a fail-fast check which asserts machine CIDR is required.

ironcladlou commented:
I still really owe this work some unit tests, but it's ready for some detailed review. So far testing has been manual.

ironcladlou commented:
cc @celebdor would love any feedback and fact checking you can spare on this one

hexfusion commented on pkg/cmd/render/bootstrap_ip_linux.go:
```diff
@@ -0,0 +1,67 @@
+	if err != nil {
+		return nil, err
+	}
+	defer nlHandle.Delete()
```
any concerns here about the route itself are we certain this is safe?

ironcladlou replied:
It's just an ambiguous API — `Close()` is probably a better name. It releases sockets/fds.

ironcladlou commented on pkg/cmd/render/bootstrap_ip.go:
```diff
@@ -0,0 +1,199 @@
+// further filter down which addresses are considered.
+//
+// This is ported from https://github.com/openshift/baremetal-runtimecfg/blob/master/pkg/utils/utils.go.
+func routableAddresses(addrMap addrMap, routeMap routeMap, vips []net.IP, af AddressFilter, rf RouteFilter) ([]net.IP, error) {
```
>I think it is in all of our best interests if this were part of some shared library. at a min we should TODO that. I forsee a bug they patch and we don't etc.

It would be useful to enumerate every use case for it outside this repo to justify doing so.

>Do we have any perf concerns with all of these nested loops I mean I guess its a result of the complexity of the problem. But I am curious what render looks like on a box with a bunch of interfaces. Does the kernel or something else do this work that we can ask?

On a box with 6 interfaces the command completes apparently "instantaneously" although I don't have the wall clock measurement from the box I was using which is long gone. I can't really conceive of any performance concerns in this context.

hexfusion replied:
thanks for looking

ironcladlou commented on pkg/cmd/render/bootstrap_ip_linux.go:
```diff
@@ -0,0 +1,67 @@
+package render
```
Probably not, but I was trying to be a good idiomatic Go citizen and ensure that the codebase will still compile on non-linux platforms (e.g. darwin). Without architecture build tags for this the operator itself will no longer compile on macos because of the native netlink dependency.

ironcladlou commented on pkg/cmd/render/render.go:
```diff
@@ -86,7 +128,9 @@ func (r *renderOpts) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&r.setupEtcdEnvImage, "manifest-setup-etcd-env-image", r.setupEtcdEnvImage, "setup-etcd-env manifest image")
 	fs.StringVar(&r.etcdDiscoveryDomain, "etcd-discovery-domain", r.etcdDiscoveryDomain, "etcd discovery domain")
-	fs.StringVar(&r.clusterConfigFile, "cluster-config-file", r.clusterConfigFile, "Openshift Cluster API Config file.")
+	// TODO: This flag name needs changed to be less confusing.
```
I would have changed it but that would would require a migration, so figured I'd note it for now

ironcladlou commented on pkg/cmd/render/bootstrap_ip.go:
```diff
@@ -0,0 +1,199 @@
+// RouteFilter is a function type to filter routes
+type RouteFilter func(netlink.Route) bool
+
+// NonDeprecatedAddress returns true if the address is IPv6 and has a preferred lifetime of 0
```
Pretty sure it's to ensure we ignore addresses like these: https://en.wikipedia.org/wiki/IPv6_address#Deprecated_and_obsolete_addresses_2

I spelled this out a little more clearly in the commit message but do need to fix up the code comments as well. Assuming my understanding and reference there is accurate.

ironcladlou replied:
https://tools.ietf.org/html/rfc4862

```
 preferred lifetime -  the length of time that a valid address is
      preferred (i.e., the time until deprecation).  When the preferred
      lifetime expires, the address becomes deprecated.

...

 IPv6 addresses are leased to an interface for a fixed (possibly
   infinite) length of time.  Each address has an associated lifetime
   that indicates how long the address is bound to an interface.  When a
   lifetime expires, the binding (and address) become invalid and the
   address may be reassigned to another interface elsewhere in the
   Internet.  To handle the expiration of address bindings gracefully,
   an address goes through two distinct phases while assigned to an
   interface.  Initially, an address is "preferred", meaning that its
   use in arbitrary communication is unrestricted.  Later, an address
   becomes "deprecated" in anticipation that its current interface
   binding will become invalid.  While an address is in a deprecated
   state, its use is discouraged, but not strictly forbidden.  New
   communication (e.g., the opening of a new TCP connection) should use
   a preferred address when possible.  A deprecated address should be
   used only by applications that have been using it and would have
   difficulty switching to another address without a service disruption.
```

I wonder if that's a better description of "why"

ironcladlou commented on pkg/cmd/render/bootstrap_ip.go:
```diff
@@ -0,0 +1,199 @@
+	return route.Dst != nil
+}
+
+func ContainedByCIDR(cidr string) AddressFilter {
```
You might want to take a closer look; this is a function generator. Not sure exactly what you're proposing

hexfusion replied:
yup disregard, thought I saw`AddressFilter` as a type bool.

ironcladlou commented on pkg/cmd/render/bootstrap_ip.go:
```diff
@@ -0,0 +1,199 @@
+		ContainedByCIDR(machineCIDR),
+		AddressNotIn(excludedIPs...),
+	)
+	routeFilter := NonDefaultRoute
```
The local variable is useless in this case, will remove it

ironcladlou commented:
Added a couple of unit tests. During the course of that I had to do a little refactoring and also discovered a bit of brittle logic around ipv6 handling (using length check to detect the family can be flaky depending on how the address struct is allocated; I switched to using `net.To4()` which I think is more stable).

@hexfusion ptal

hexfusion commented:
/retest

hexfusion commented:
quick check of aws[1] failure etcd bootstrapped fine.

[1] https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/384/pull-ci-openshift-cluster-etcd-operator-master-e2e-aws/1280539049492221952
/retest

hexfusion commented:
hmm CI looks hosed :(

hexfusion commented:
/retest

hexfusion commented:
/retest

hexfusion commented:
/test e2e-gcp-upgrade

hexfusion commented:
/test e2e-gcp-upgrade

hexfusion commented:
/test e2e-gcp-upgrade

ironcladlou commented:
The metal failure could be something legit with ipv6, need to investigate

hexfusion commented:
@stbenjam we are nervous to merge this with metal failing.

cc @romfreiman

ironcladlou commented:
Okay, I had a chance to dig in to the metal-ipi failures, and the bootstrap IP seems to have been successfully identified and etcd is coming up — I didn't dig further into what might be the root cause of the overall failures. I need to rebase now, so I'll do that and squash again and go through another round of CI tests.

hexfusion commented:
/lgtm
/approve
